File: server/src/Makc2021.Layer5.Apps.WebAPI/Pages/DummyMain/List/DummyMainListPageService.cs
﻿// Copyright (c) 2021 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Makc2021.Layer1.Completion;
using Makc2021.Layer1.Query;
using Makc2021.Layer4.Domains.DummyMain;
using Makc2021.Layer4.Domains.DummyMain.Queries.List.Get;
using Makc2021.Layer5.Apps.WebAPI.Pages.DummyMain.List.Queries.Get;

namespace Makc2021.Layer5.Apps.WebAPI.Pages.DummyMain.List
{
    /// <summary>
    /// Сервис страницы сущностей "DummyMain".
    /// </summary>
    public class DummyMainListPageService : IDummyMainListPageService
    {
        private IListGetQueryDomainHandler AppListGetQueryDomainHandler { get; }

        private IDomainService AppService { get; }

        #region Constructors

        /// <summary>
        /// Конструктор.
        /// </summary>
        /// <param name="appListGetQueryDomainHandler">Обработчик запроса на получение.</param>
        /// <param name="appService">Сервис.</param>
        public DummyMainListPageService(
            IListGetQueryDomainHandler appListGetQueryDomainHandler,
            IDomainService appService
            )
        {
            AppListGetQueryDomainHandler = appListGetQueryDomainHandler;
            AppService = appService;
        }

        #endregion Constructors

        #region Public methods

        /// <inheritdoc/>
        public async Task<QueryResultWithOutput<DummyMainListPageGetQueryOutput>> Get(DummyMainListPageGetQueryInput input)
        {
            QueryResultWithOutput<DummyMainListPageGetQueryOutput> result = new();

            DummyMainListPageGetQueryOutput output = new();

            List<QueryResult> queryResults = new();

            var queryResult1 = await  GetListGetQueryResult(
                result.QueryCode,
                new ListGetQueryDomainInput
                {
                    PageNumber = input.List.PageNumber,
                    PageSize = input.List.PageSize
                })
                .ConfigureAwaitWithCultureSaving(false);

            queryResults.Add(queryResult1);

            if (queryResult1.IsOk && queryResult1.Output != null)
            {
                output.List = queryResult1.Output;
            }

            result.Load(queryResults);

            if (result.IsOk)
            {
                result.Output = output;
            }

            return result;
        }

        #endregion Public methods

        #region Private methods

        private async Task<QueryResultWithOutput<ListGetQueryDomainOutput>> GetListGetQueryResult(
            string queryCode,
            ListGetQueryDomainInput input
            )
        {
            var queryHandler = AppListGetQueryDomainHandler;

            queryHandler.QueryResult.QueryCode = queryCode;

            try
            {
                queryHandler.OnStart(input);

                var queryOutput = await AppService.GetList(queryHandler.QueryInput)
                    .ConfigureAwaitWithCultureSaving(false);

                queryHandler.OnSuccess(queryOutput);
            }
            catch (Exception ex)
            {
                queryHandler.OnError(ex);
            }

            return queryHandler.QueryResult;
        }

        #endregion Private methods
    }
}

File: client/makc2021-react-redux-ts/src/Layer5/Pages/DummyMain/List/DummyMainListPageService.ts
// Copyright (c) 2021 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

import { QueryResultWithOutput } from 'src/Layer1/Query/QueryResultWithOutput';
import { DummyMainListPageGetQueryInput } from './Queries/Get/DummyMainListPageGetQueryInput';
import { DummyMainListPageGetQueryOutput } from './Queries/Get/DummyMainListPageGetQueryOutput';
import { DummyMainListPageGetQueryHandler } from './Queries/Get/DummyMainListPageGetQueryHandler';

/**
 * Сервис страницы сущностей "DummyMain".
 */
export class DummyMainListPageService {
  /**
   * Конструктор.
   * @param _appService Сервис.
   * @param _appUrlService Сервис URL.
   * @param _appGetQueryHandler Обработчик запроса на получение.
   */
  constructor(private _appGetQueryHandler: DummyMainListPageGetQueryHandler) {}

  /**
   * Получить.
   * @param input Входные данные.
   * @returns Обещание получения результата с выходными данными.
   */
  async get(
    input: DummyMainListPageGetQueryInput
  ): Promise<QueryResultWithOutput<DummyMainListPageGetQueryOutput>> {
    const queryHandler = this._appGetQueryHandler;

    try {
      queryHandler.onStart(input);

      const queryResult = await queryHandler.executeQuery();

      queryHandler.onSuccess(queryResult);
    } catch (error) {
      queryHandler.onError(error);
    }

    return queryHandler.queryResult;
  }
}
